### ISSUE TITLE
YCbCr colorspace from pyvips jp2k/jpeg compression causes brightfield/RGB ome.tiff image results to look pink when opening in QuPath

### ISSUE THREAD

**Issue by snibbor** [MathOnco/valis#26]
Hi Chandler,

The YCbCr colorspace from pyvips jp2k/jpeg compression causes brightfield/RGB ome.tiff image results to look pink when opening in QuPath:

![YCbCr QuPath Pink](https://user-images.githubusercontent.com/28576964/198887068-9c85e188-fbdf-4c7c-bf7d-7bb425de0810.png)

So this is a minor problem if you want to use QuPath with these results. The easy fix is to add an alpha channel if you are saving a RGB/brightfield image. Like this:

```
max_alpha = 255
output = output.bandjoin(max_alpha)
output = output.copy(interpretation="srgb")
ome_tiff_name = os.path.split(ome_tiff_path)[-1]
new_name = ome_tiff_name.split(".ome.tiff")[0] + "DABpCol.ome.tiff"
print("saving DAB pseudo-colored image... \n{} in {}".format(new_name, output_dir))
image_height = output.height
image_bands = output.bands

# split to separate image planes and stack vertically ready for OME, not needed for sRGB or RGB
# output = pyvips.Image.arrayjoin(output.bandsplit(), across=1)

# set minimal OME metadata
# before we can modify an image (set metadata in this case), we must take a
# private copy
output = output.copy()
output.set_type(pyvips.GValue.gint_type, "page-height", image_height)
output.set_type(pyvips.GValue.gstr_type, "image-description",
	              f"""<?xml version="1.0" encoding="UTF-8"?>
			  <OME xmlns="http://www.openmicroscopy.org/Schemas/OME/2016-06"
			      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
			      xsi:schemaLocation="http://www.openmicroscopy.org/Schemas/OME/2016-06 http://www.openmicroscopy.org/Schemas/OME/2016-06/ome.xsd">
			  <Instrument ID="Instrument:0">
			      <Objective ID="Objective:0" NominalMagnification="20.0"/>
		          </Instrument>
			  <Image ID="Image:0">
			      <InstrumentRef ID="Instrument:0"/>
			      <ObjectiveSettings ID="Objective:0"/>
			      <Pixels DimensionOrder="XYCZT"
			              ID="Pixels:0"
			              PhysicalSizeX="0.325"
			              PhysicalSizeXUnit="µm" 
			              PhysicalSizeY="0.325" 
			              PhysicalSizeYUnit="µm" 
			              PhysicalSizeZ="0.7787" 
			              PhysicalSizeZUnit="µm" 
			              SizeC="{image_bands}"
			              SizeT="1"
			              SizeX="{output.width}"
			              SizeY="{image_height}"
			              SizeZ="1"
			              Type="uint8">
			              <Channel ID="Channel:0:0" SamplesPerPixel="3">
			                  <LightPath/>
		                  </Channel>
			      </Pixels>
			  </Image>
	</OME>""")

output.tiffsave(os.path.join(output_dir, new_name),
	              compression="jp2k",
	              Q=100,
	              lossless=True,
	              tile=True, tile_width=1024, tile_height=1024,
	              pyramid=True, subifd=True,
	              bigtiff=True
	              )
```

I patched my VALIS fork this way since I am using these registered images with QuPath, but this might not be the case for everyone and the YCbCr colorspace gives better compression with jp2k/jpeg. It's more of a bug with QuPath than VALIS.

Just wanted to post my solution here for others to find, but not sure if VALIS needs this fix or not. 

Best,
Jack

**Comment by cdgatenbee** [comment]
Hi @crobbins327 ,
This is interesting, thanks for posting this. I hadn't realized that libvips encodes RGB images as YCbCr when coding for jpeg (https://legacy.imagemagick.org/discourse-server/viewtopic.php?t=32729). I'd like to experiment a bit, but maybe it would be good to incorporate your fix whenever compression options are set to jpeg/jp2k and the image is RGB. Also, I did want to let you know that I've addressed the issues about checking for mirroring and keeping the image's original colors. They'll be included in the next update, which I should be able to push early next week.  

Best,
Chandler

**Comment by crobbins327** [comment]
Awesome Chandler @cdgatenbee! Looking forward to it.

Please let me know if you find an option to just directly save as photometric RGB instead of YCbCr in pyvips. Adding the alpha channel is kinda hacky, rather not have to do that when compressing using jp2k/jpeg.

**Comment by cdgatenbee** [comment]
Hi @snibbor,
I recently found a solution to this issue, which was setting ```rgbjpeg=True``` in pyvips' ```Image.tiffsave``` function. In version 1.0.1, I added some code to valis' ```slide_io.save_ome_tiff``` that sets ```rgbjpeg=True``` if the image is rgb and the compression is "jpeg" or "jp2k", and now the compressed rgb images look as expected when opening in QuPath. I believe that resolves this issue, so I'll go ahead and close, but let me know if you run into any problems.
Best,
-Chandler
